File: UnrealEngine/Engine/Source/Developer/Horde/Private/Storage/Bundles/V2/PacketWriter.cpp
// Copyright Epic Games, Inc. All Rights Reserved.

#include "Storage/Bundles/V2/PacketWriter.h"
#include "Storage/Bundles/Bundle.h"
#include "Storage/BlobWriter.h"
#include "Memory/SharedBuffer.h"
#include "../../../HordePlatform.h"

FPacketWriter::FPacketWriter(FBlobHandle InBundleHandle, FBlobHandle InPacketHandle)
	: BundleHandle(MoveTemp(InBundleHandle))
	, PacketHandle(MoveTemp(InPacketHandle))
	, Buffer(FUniqueBuffer::Alloc(1024).MoveToShared())
{
	ImportHandles.SetNum(FPacketImport::Bias);
	ImportHandles[FPacketImport::CurrentBundleBaseIdx + FPacketImport::Bias] = BundleHandle;
	ImportHandles[FPacketImport::CurrentPacketBaseIdx + FPacketImport::Bias] = PacketHandle;

	ImportMap.Add(BundleHandle, FPacketImport::CurrentBundleBaseIdx);
	ImportMap.Add(PacketHandle, FPacketImport::CurrentPacketBaseIdx);

	ImportOffsets.Add(0);

	BufferBasePtr = (uint8*)Buffer.GetData();
	FBundleSignature(EBundleVersion::LatestV2, 0).Write(BufferBasePtr);

	Length = FBundleSignature::NumBytes + (sizeof(int) * 3);
	ExportOffsets.Add((uint32)Length);

	NextBlobLength = 0;// sizeof(int32);
}

FPacketWriter::~FPacketWriter()
{
}

size_t FPacketWriter::GetLength() const
{
	return Length + NextBlobLength;
}

void FPacketWriter::AddImport(FBlobHandle Import)
{
	int32 ImportIdx = FindOrAddImport(Import);
	NextBlobImports.Add(ImportIdx);
}

int FPacketWriter::GetImportCount() const
{
	return ImportHandles.Num();
}

FBlobHandle FPacketWriter::GetImport(int ImportIdx) const
{
	return ImportHandles[ImportIdx];
}

FMutableMemoryView FPacketWriter::GetOutputBuffer(size_t UsedSize, size_t DesiredSize)
{
	size_t TotalUsedSize = NextBlobLength + UsedSize;
	if (TotalUsedSize > 0)
	{
		TotalUsedSize += sizeof(int32);
	}

	FMutableMemoryView Memory = GetOutputBufferInternal(TotalUsedSize, sizeof(int32) + NextBlobLength + DesiredSize);
	return Memory.Mid(sizeof(int32) + NextBlobLength);
}

void FPacketWriter::Advance(size_t Size)
{
	NextBlobLength += Size;
}

int FPacketWriter::CompleteBlob(const FBlobType& Type)
{
	size_t TypeIdx = FindOrAddType(Type);

	// Ensure the buffer has been allocated
	GetOutputBuffer(0, 0);

	// Finalize the written export data, inserting the length at the start.
	*(int*)(BufferBasePtr + Length) = (int)NextBlobLength;
	Length += sizeof(int) + NextBlobLength;

	// Write the type and import list
	uint8* Span = (uint8*)GetOutputBufferInternal(0, (NextBlobImports.Num() + 2) * 5).GetData();
	size_t Offset = 0;

	// Write the type and import list
	Offset += WriteUnsignedVarInt(Span + Offset, TypeIdx);
	Offset += WriteUnsignedVarInt(Span + Offset, NextBlobImports.Num());

	for (int32 Import : NextBlobImports)
	{
		Offset += WriteUnsignedVarInt(Span + Offset, Import);
	}

	Length = Align(Length + Offset);

	// Reset the current export
	NextBlobLength = 0;
	NextBlobImports.Empty();

	// Write the next export offset to the buffer
	ExportOffsets.Add((uint32)Length);
	return ExportOffsets.Num() - 2;
}

int FPacketWriter::GetExportCount() const
{
	return ExportOffsets.Num() - 1;
}

FBlob FPacketWriter::GetExport(int ExportIdx) const
{
	size_t PacketOffset = ExportOffsets[ExportIdx];
	size_t PacketLength = ExportOffsets[ExportIdx + 1] - PacketOffset;

	FPacketExport Export(FSharedBufferView(Buffer, PacketOffset, PacketLength));

	FBlobType Type = Types[Export.GetTypeIndex()];

	TArray<size_t> ImportIndices;
	Export.GetImportIndices(ImportIndices);

	TArray<FBlobHandle> Imports;
	for (size_t ImportIdx : ImportIndices)
	{
		Imports.Add(ImportHandles[ImportIdx + FPacketImport::Bias]);
	}

	return FBlob(Type, Export.GetPayload(), MoveTemp(Imports));
}

FPacket FPacketWriter::CompletePacket()
{
	Length = Align(Length);

	// Write the various lookup tables
	WriteTypeTable();
	WriteImportTable();
	WriteExportTable();

	// Write the final packet length
	FBundleSignature(EBundleVersion::LatestV2, Length).Write(BufferBasePtr);
	return FPacket(FSharedBufferView(Buffer, 0, Length));
}

FMutableMemoryView FPacketWriter::GetOutputBufferInternal(size_t UsedSize, size_t DesiredSize)
{
	if (Length + DesiredSize > Buffer.GetSize())
	{
		int NewSize = (Length + DesiredSize + 4096 + 16384) & ~16384;

		FSharedBuffer NewBuffer = FUniqueBuffer::Alloc(NewSize).MoveToShared();
		uint8* NewBufferBasePtr = (uint8*)NewBuffer.GetData();
		memcpy(NewBufferBasePtr, BufferBasePtr, Length + UsedSize);

		Buffer = MoveTemp(NewBuffer);
		BufferBasePtr = NewBufferBasePtr;
	}
	return FMutableMemoryView(BufferBasePtr + Length, Buffer.GetSize() - Length);
}

unsigned char* FPacketWriter::GetOutputSpanAndAdvance(size_t Size)
{
	FMutableMemoryView Span = GetOutputBufferInternal(0, Size);
	Length += Size;
	return (uint8*)Span.GetData();
}

int32 FPacketWriter::FindOrAddType(FBlobType Type)
{
	for (int TypeIdx = 0; TypeIdx < Types.Num(); TypeIdx++)
	{
		if (Types[TypeIdx] == Type)
		{
			return TypeIdx;
		}
	}

	Types.Add(Type);
	return Types.Num() - 1;
}

int32 FPacketWriter::FindOrAddImport(const FBlobHandle& Handle)
{
	int32* ImportIdxPtr = ImportMap.Find(Handle);
	if (ImportIdxPtr == nullptr)
	{
		FUtf8String Fragment;
		verify(Handle->TryAppendIdentifier(Fragment));

		FBlobHandle Outer = Handle->GetOuter();

		int BaseIdx;
		if (Outer)
		{
			BaseIdx = FindOrAddImport(Outer);
		}
		else
		{
			BaseIdx = FPacketImport::InvalidBaseIdx;
		}

		size_t BaseIdxLen = MeasureUnsignedVarInt(BaseIdx + FPacketImport::Bias);
		size_t ImportLen = BaseIdxLen + Fragment.Len();

		FMutableMemoryView ImportData = ImportWriter.GetOutputBuffer(0, ImportLen);
		WriteUnsignedVarIntWithKnownLength(ImportData.GetData(), BaseIdx + FPacketImport::Bias, BaseIdxLen);
		memcpy((char*)ImportData.GetData() + BaseIdxLen, *Fragment, Fragment.Len());
		ImportWriter.Advance(ImportLen);

		int32 ImportIdx = ImportOffsets.Num() - 1;
		ImportHandles.Add(Handle);
		ImportOffsets.Add(ImportWriter.GetLength());
		ImportIdxPtr = &ImportMap.Add(Handle, ImportIdx);
	}
	return *ImportIdxPtr;
}

void FPacketWriter::WriteTypeTable()
{
	// Write the type table offset in the header
	*(int*)(BufferBasePtr + 8) = (int)Length;

	// Write the types to the end of the buffer
	unsigned char* Span = GetOutputSpanAndAdvance(sizeof(int) + Types.Num() * sizeof(FBlobType));

	*(int*)Span = (int)Types.Num();
	Span += sizeof(int);

	memcpy(Span, Types.GetData(), Types.Num() * sizeof(FBlobType));
}

void FPacketWriter::WriteImportTable()
{
	// Write the the import table offset in the header
	*(int*)(BufferBasePtr + 12) = (int)Length;

	// Allocate the buffers
	int ImportHeaderLength = (ImportOffsets.Num() + 1) * sizeof(int);
	int ImportDataOffset = Length + ImportHeaderLength;
	unsigned char* Span = GetOutputSpanAndAdvance(Align(ImportHeaderLength + ImportWriter.GetLength()));

	// Write the import header
	*(int*)Span = ImportOffsets.Num() - 1;
	Span += sizeof(int);

	for (int ImportOffset : ImportOffsets)
	{
		*(int*)Span = ImportDataOffset + ImportOffset;
		Span += sizeof(int);
	}

	// Write the import data
	ImportWriter.CopyTo(Span);
}

void FPacketWriter::WriteExportTable()
{
	// Write the export table offset in the header
	*(int*)(BufferBasePtr + 16) = (int)Length;

	// Write the exports
	int* Span = (int*)GetOutputSpanAndAdvance((ExportOffsets.Num() + 1) * sizeof(int));
	*(Span++) = (int)(ExportOffsets.Num() - 1);
	memcpy(Span, ExportOffsets.GetData(), ExportOffsets.Num() * sizeof(int));
}

size_t FPacketWriter::Align(size_t Value)
{
	return (Value + 3) & ~3;
}

File: UnrealEngine/Engine/Source/Programs/Shared/EpicGames.Horde/Storage/Bundles/V2/PacketWriter.cs
// Copyright Epic Games, Inc. All Rights Reserved.

using System;
using System.Buffers.Binary;
using System.Collections.Generic;
using System.Diagnostics;
using System.Runtime.InteropServices;
using EpicGames.Core;

namespace EpicGames.Horde.Storage.Bundles.V2
{
	/// <summary>
	/// Writes exports into a new bundle packet
	/// </summary>
	public sealed class PacketWriter : IDisposable
	{
		const string AllocationTag = nameof(PacketWriter);

		readonly BundleHandle _bundleHandle;
		readonly PacketHandle _packetHandle;
		readonly IMemoryAllocator<byte> _allocator;
		readonly object _lockObject;

		IRefCountedHandle<Memory<byte>> _bufferHandle;
		Memory<byte> _buffer;

		int _length;

		readonly List<BlobType> _types = new List<BlobType>();
		readonly List<PacketImport> _imports = new List<PacketImport>();
		readonly List<object> _importHandles = new List<object>();
		readonly Dictionary<object, int> _importMap = new Dictionary<object, int>();
		readonly List<int> _exportOffsets = new List<int>();

		/// <summary>
		/// Current length of the packet
		/// </summary>
		public int Length => _length;

		/// <summary>
		/// Constructor
		/// </summary>
		public PacketWriter(BundleHandle bundleHandle, PacketHandle packetHandle, IMemoryAllocator<byte> allocator, object lockObject)
		{
			_bundleHandle = bundleHandle;
			_packetHandle = packetHandle;
			_allocator = allocator;
			_lockObject = lockObject;

			for (int idx = 0; idx < PacketImport.Bias; idx++)
			{
				_importHandles.Add(null!);
			}

			_importHandles[PacketImport.CurrentBundleBaseIdx + PacketImport.Bias] = _bundleHandle;
			_importHandles[PacketImport.CurrentPacketBaseIdx + PacketImport.Bias] = _packetHandle;

			_importMap[_bundleHandle] = PacketImport.CurrentBundleBaseIdx;
			_importMap[_packetHandle] = PacketImport.CurrentPacketBaseIdx;

			_bufferHandle = RefCountedHandle.Create(_allocator.Alloc(1024, AllocationTag));
			_buffer = _bufferHandle.Target;

			BundleSignature signature = new BundleSignature(BundleVersion.LatestV2, 0);
			signature.Write(_buffer.Span);

			_length = BundleSignature.NumBytes + (sizeof(int) * 3);
			_exportOffsets.Add(_length);
		}

		/// <inheritdoc/>
		public void Dispose()
		{
			_buffer = Memory<byte>.Empty;

			if (_bufferHandle != null)
			{
				_bufferHandle.Dispose();
				_bufferHandle = null!;
			}
		}

		/// <summary>
		/// Gets the number of exports currently in this writer
		/// </summary>
		public int GetExportCount() => _exportOffsets.Count - 1;

		/// <summary>
		/// Writes a new blob to this packet
		/// </summary>
		/// <param name="size"></param>
		/// <param name="type"></param>
		/// <param name="imports"></param>
		public int CompleteExport(int size, BlobType type, IReadOnlyList<IBlobHandle> imports)
		{
			int[] importIndices = new int[imports.Count];
			for (int idx = 0; idx < imports.Count; idx++)
			{
				importIndices[idx] = FindOrAddImport(imports[idx].Innermost);
			}

			int typeIdx = FindOrAddType(type);
			return CompleteExport(size, typeIdx, importIndices);
		}

		/// <summary>
		/// Complete the current export
		/// </summary>
		/// <param name="size">Size of data written to the export buffer</param>
		/// <param name="typeIdx">Index of the blob type</param>
		/// <param name="imports">Indexes of the imported blobs</param>
		internal int CompleteExport(int size, int typeIdx, int[] imports)
		{
			// Finalize the written export data, inserting the length at the start.
			BinaryPrimitives.WriteInt32LittleEndian(_buffer.Slice(_length).Span, size);
			_length += sizeof(int) + size;

			// Write the type and import list
			Span<byte> span = GetOutputBufferInternal(0, (imports.Length + 2) * 5).Span;
			int offset = 0;

			offset += VarInt.WriteUnsigned(span[offset..], typeIdx);
			offset += VarInt.WriteUnsigned(span[offset..], imports.Length);

			for (int idx = 0; idx < imports.Length; idx++)
			{
				offset += VarInt.WriteUnsigned(span[offset..], imports[idx]);
			}

			_length = Align(_length + offset, 4);

			// Write the next export offset to the buffer
			_exportOffsets.Add(_length);
			return _exportOffsets.Count - 2;
		}

		/// <summary>
		/// Reads data for a blob written to storage
		/// </summary>
		/// <param name="exportIdx"></param>
		/// <returns></returns>
		public BlobData GetExport(int exportIdx)
		{
			lock (_lockObject)
			{
				int offset = _exportOffsets[exportIdx];
				int length = _exportOffsets[exportIdx + 1] - offset;

				PacketExport export = new PacketExport(_buffer.Slice(offset, length));
				PacketExportHeader header = export.GetHeader();

				BlobType type = _types[header.TypeIdx];

				IBlobHandle[] imports = new IBlobHandle[header.Imports.Length];
				for (int idx = 0; idx < header.Imports.Length; idx++)
				{
					int importIdx = header.Imports[idx];
					imports[idx] = (IBlobHandle)_importHandles[importIdx + PacketImport.Bias];
				}

				IReadOnlyMemoryOwner<byte> body = ReadOnlyMemoryOwner.Create(export.GetPayload(), _bufferHandle.AddRef());
				return new BlobDataWithOwner(_types[header.TypeIdx], body.Memory, imports, body);
			}
		}

		/// <summary>
		/// Add a new blob type to be written
		/// </summary>
		/// <param name="type">Type to add</param>
		/// <returns>Index of the type</returns>
		public int FindOrAddType(BlobType type)
		{
			int idx = _types.IndexOf(type);
			if (idx == -1)
			{
				idx = _types.Count;
				_types.Add(type);
			}
			return idx;
		}

		int AddImport(int baseIdx, Utf8String fragment, object handle)
		{
			PacketImport import = new PacketImport(baseIdx, fragment);
			int idx = _imports.Count;
			_imports.Add(import);
			_importMap.Add(handle, idx);

			Debug.Assert(idx + PacketImport.Bias == _importHandles.Count);
			_importHandles.Add(handle);

			return idx;
		}

		/// <summary>
		/// Adds a new imported blob locator
		/// </summary>
		/// <param name="handle">Handle to add</param>
		/// <returns>Index of the import</returns>
		public int FindOrAddImport(IBlobHandle handle) => FindOrAddImportInternal(handle);

		int FindOrAddImportInternal(object handle)
		{
			int idx;
			if (_importMap.TryGetValue(handle, out idx))
			{
				return idx;
			}
			else if (handle is ExportHandle exportHandle)
			{
				int baseIdx = FindOrAddImportInternal(exportHandle.Packet);
				return AddImport(baseIdx, exportHandle.GetIdentifier(), handle);
			}
			else if (handle is PacketHandle packetHandle)
			{
				Utf8StringBuilder builder = new Utf8StringBuilder();
				if (!packetHandle.TryAppendIdentifier(builder))
				{
					throw new NotSupportedException("Referenced blob has not been flushed to storage yet");
				}

				int baseIdx = FindOrAddImportInternal(packetHandle.Bundle);
				return AddImport(baseIdx, builder.ToUtf8String(), handle);
			}
			else if (handle is BundleHandle bundleHandle)
			{
				BlobLocator locator;
				if (!bundleHandle.TryGetLocator(out locator))
				{
					throw new NotSupportedException("Referenced blob has not been flushed to storage yet");
				}
				return AddImport(-1, locator.Path, handle);
			}
			else if (handle is IBlobHandle blobHandle)
			{
				BlobLocator locator;
				if (!blobHandle.TryGetLocator(out locator))
				{
					throw new NotSupportedException("Referenced blob has not been flushed to storage yet");
				}
				return AddImport(-1, locator.Path, handle);
			}
			else
			{
				throw new NotSupportedException("Unknown blob type");
			}
		}

		/// <summary>
		/// Gets the import assigned to a particular index
		/// </summary>
		public IBlobHandle GetImport(int importIdx)
			=> (_importHandles[importIdx + PacketImport.Bias] as IBlobHandle) ?? throw new InvalidOperationException("Import is not a blob handle");

		/// <summary>
		/// Gets data to write new export
		/// </summary>
		/// <param name="usedSize">Size of data in the current buffer that has been written to</param>
		/// <param name="desiredSize"></param>
		/// <returns></returns>
		public Memory<byte> GetOutputBuffer(int usedSize, int desiredSize)
		{
			int totalUsedSize = 0;
			if (usedSize > 0)
			{
				totalUsedSize = sizeof(int) + usedSize;
			}

			Memory<byte> memory = GetOutputBufferInternal(totalUsedSize, sizeof(int) + desiredSize);
			return memory[4..];
		}

		/// <summary>
		/// Aligns an offset to a power-of-2 boundary
		/// </summary>
		static int Align(int value, int alignment) => (value + (alignment - 1)) & ~(alignment - 1);

		/// <summary>
		/// Gets data to write new export
		/// </summary>
		Memory<byte> GetOutputBufferInternal(int usedSize, int desiredSize)
		{
			if (_length + desiredSize > _buffer.Length)
			{
				lock (_lockObject)
				{
					int newSize = (_length + desiredSize + 4096 + 16384) & ~16384;

					IRefCountedHandle<Memory<byte>> newBufferHandle = RefCountedHandle.Create(_allocator.Alloc(newSize, AllocationTag));
					_buffer.Slice(0, _length + usedSize).CopyTo(newBufferHandle.Target);
					_buffer = newBufferHandle.Target;
					_bufferHandle.Dispose();

					_bufferHandle = newBufferHandle;
				}
			}
			return _buffer.Slice(_length);
		}

		/// <summary>
		/// Gets an output span and updates the length accordingly
		/// </summary>
		Span<byte> GetOutputSpanAndAdvance(int length)
		{
			Span<byte> span = GetOutputBufferInternal(0, length).Span;
			_length += length;
			return span.Slice(0, length);
		}

		/// <summary>
		/// Mark the current packet as complete
		/// </summary>
		public Packet CompletePacket()
		{
			_length = Align(_length, 4);

			// Write the various lookup tables
			WriteTypeTable();
			WriteImportTable();
			WriteExportTable();

			// Write the final packet length
			BinaryPrimitives.WriteInt32LittleEndian(_buffer.Span.Slice(4), _length);
			return new Packet(_buffer.Slice(0, _length));
		}

		void WriteTypeTable()
		{
			// Write the type table offset in the header
			BinaryPrimitives.WriteInt32LittleEndian(_buffer.Span.Slice(8), _length);

			// Write the types to the end of the buffer
			Span<byte> span = GetOutputSpanAndAdvance(sizeof(int) + _types.Count * BlobType.NumBytes);

			BinaryPrimitives.WriteInt32LittleEndian(span, _types.Count);
			span = span[4..];

			foreach (BlobType type in _types)
			{
				type.Write(span);
				span = span[IoHash.NumBytes..];
			}
		}

		void WriteImportTable()
		{
			// Write the the import table offset in the header
			BinaryPrimitives.WriteInt32LittleEndian(_buffer.Span.Slice(12), _length);

			// Measure the length of all the imports
			int importDataLength = 0;
			foreach (PacketImport import in _imports)
			{
				importDataLength += VarInt.MeasureUnsigned(import.BaseIdx + PacketImport.Bias) + import.Fragment.Length;
			}
			importDataLength = Align(importDataLength, 4);

			// Allocate the buffers
			int importHeaderLength = (_imports.Count + 2) * sizeof(int);
			int importDataOffset = _length + importHeaderLength;
			Span<byte> span = GetOutputSpanAndAdvance(importHeaderLength + importDataLength);

			// Write all the import data
			BinaryPrimitives.WriteInt32LittleEndian(span, _imports.Count);
			span = span[4..];

			foreach (PacketImport import in _imports)
			{
				BinaryPrimitives.WriteInt32LittleEndian(span, importDataOffset);
				span = span[4..];

				Span<byte> baseSpan = _buffer.Span.Slice(importDataOffset);
				importDataOffset += VarInt.WriteUnsigned(baseSpan, import.BaseIdx + PacketImport.Bias);

				Span<byte> fragmentSpan = _buffer.Span.Slice(importDataOffset);
				import.Fragment.Span.CopyTo(fragmentSpan);
				importDataOffset += import.Fragment.Length;
			}
			BinaryPrimitives.WriteInt32LittleEndian(span, importDataOffset);
		}

		void WriteExportTable()
		{
			// Write the export table offset in the header
			BinaryPrimitives.WriteInt32LittleEndian(_buffer.Span.Slice(16), _length);

			// Write the exports
			Span<int> span = MemoryMarshal.Cast<byte, int>(GetOutputSpanAndAdvance((_exportOffsets.Count + 1) * sizeof(int)));

			span[0] = _exportOffsets.Count - 1;
			for (int idx = 0; idx < _exportOffsets.Count; idx++)
			{
				span[idx + 1] = _exportOffsets[idx];
			}
		}
	}
}
